File: scripts/decode_chunk.py
#!/usr/bin/env python3
"""Utility to decode a single chunk from ``Obfuscated.json``.

The helper mirrors the initv4 bootstrapper pipeline as closely as possible:

* Collect the chunk into a single printable blob.
* Attempt to recover the initv4 alphabet from the bootstrapper (if available)
  and decode the blob using the initv4 basE91 variant.
* Fall back to the default alphabet or an ascii85 decode when basE91 fails.
* Apply the script-key XOR stage followed by the index XOR mixing.

When the final byte stream cannot be decoded as UTF-8 the script emits a
formatted Lua comment containing a hex/ASCII view of the payload instead of a
binary artefact.  This keeps the output text-based while still exposing the raw
bytes for further manual analysis.
"""

from __future__ import annotations

import argparse
import base64
import bz2
import json
import lzma
import re
import zlib
from pathlib import Path
from typing import Callable, Iterable, Optional, Sequence

DEFAULT_ALPHABET = (
    "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    "!#$%&()*+,-./:;<=>?@[]^_`{|}~\"'\\"
)

ALPHABET_PATTERN = re.compile(r"([!-~]{85,})")


def _load_json(path: Path) -> list[object]:
    with path.open("r", encoding="utf-8") as fh:
        return json.load(fh)


def _normalise_chunk(chunk: object) -> str:
    if isinstance(chunk, str):
        return chunk
    if isinstance(chunk, Sequence):
        try:
            return "".join(str(part) for part in chunk)
        except Exception as exc:  # pragma: no cover - defensive
            raise TypeError(f"Chunk cannot be concatenated: {chunk!r}") from exc
    raise TypeError(f"Unsupported chunk type: {type(chunk)!r}")


def _prepare_alphabet(candidate: Optional[str]) -> tuple[str, dict[str, int], int]:
    if candidate:
        unique = []
        seen: set[str] = set()
        for char in candidate:
            if char in seen:
                continue
            seen.add(char)
            unique.append(char)
        if len(unique) >= 85:
            alphabet = "".join(unique)
            return alphabet, {symbol: idx for idx, symbol in enumerate(alphabet)}, len(alphabet)
    alphabet = DEFAULT_ALPHABET
    return alphabet, {symbol: idx for idx, symbol in enumerate(alphabet)}, len(alphabet)


def _decode_base91(blob: str, *, alphabet: Optional[str] = None) -> bytes:
    """Decode *blob* using the initv4 basE91 alphabet."""

    alphabet, mapping, base = _prepare_alphabet(alphabet)
    value = -1
    buffer = 0
    bits = 0
    output = bytearray()

    for char in blob:
        if char.isspace():
            continue
        if char not in mapping:
            raise ValueError(f"invalid initv4 symbol: {char!r}")
        symbol = mapping[char]
        if value < 0:
            value = symbol
            continue

        value += symbol * base
        buffer |= value << bits
        if value & 0x1FFF > 88:
            bits += 13
        else:
            bits += 14

        while bits >= 8:
            output.append(buffer & 0xFF)
            buffer >>= 8
            bits -= 8

        value = -1

    if value + 1:
        buffer |= value << bits
        output.append(buffer & 0xFF)

    return bytes(output)


def _decode_ascii85(blob: str) -> bytes:
    try:
        return base64.a85decode(
            blob.encode("ascii"),
            adobe=False,
            ignorechars=b"\r\n\t ",
        )
    except Exception as exc:  # pragma: no cover - defensive
        raise ValueError(f"ascii85 decode failed: {exc}") from exc


def _apply_script_key_transform(data: bytes, key_bytes: bytes) -> bytes:
    if not data:
        return b""

    key_len = len(key_bytes)
    output = bytearray(len(data))
    for index, value in enumerate(data):
        if key_len:
            value ^= key_bytes[index % key_len]
        value ^= index & 0xFF
        output[index] = value & 0xFF
    return bytes(output)


def _iter_decoders(alphabet: Optional[str]) -> Iterable[tuple[str, Callable[[str], bytes]]]:
    yield "ascii85", _decode_ascii85
    yield "base91", lambda blob: _decode_base91(blob, alphabet=alphabet)


def decode_chunk(
    chunk: object,
    script_key: str,
    *,
    alphabet: Optional[str] = None,
) -> tuple[bytes, str, Optional[str], list[str]]:
    """Decode *chunk* with the initv4 pipeline.

    The helper now continues iterating through the configured decoders until one
    of them produces human readable text (either directly or after a secondary
    compression stage is peeled away).  When all strategies fail to recover
    UTF-8 we return the best-effort byte stream alongside the recorded notes so
    callers can render a hexdump fallback."""

    joined = _normalise_chunk(chunk)
    last_error: Optional[str] = None
    best_attempt: Optional[tuple[bytes, str, list[str]]] = None

    for name, decoder in _iter_decoders(alphabet):
        try:
            decoded = decoder(joined)
        except Exception as exc:  # pragma: no cover - decode fallback
            last_error = f"{name} decode failed: {exc}"
            continue

        output = _apply_script_key_transform(decoded, script_key.encode("utf-8"))
        text, notes = _attempt_transforms(output)
        if text is not None:
            return output, name, text, notes

        if best_attempt is None:
            best_attempt = (output, name, notes)

    if best_attempt is not None:
        decoded, strategy, notes = best_attempt
        return decoded, strategy, None, notes

    if last_error:
        raise ValueError(last_error)
    raise ValueError("no decode strategy succeeded")


def _auto_detect_alphabet(path: Optional[Path]) -> Optional[str]:
    if not path:
        return None
    text = path.read_text(encoding="utf-8", errors="ignore")
    longest: Optional[str] = None
    for candidate in ALPHABET_PATTERN.findall(text):
        if len(set(candidate)) >= 85:
            if longest is None or len(candidate) > len(longest):
                longest = candidate
    return longest


def _attempt_transforms(data: bytes) -> tuple[Optional[str], list[str]]:
    attempts: list[str] = []
    # zlib/bzip2/lzma are the most common secondary wrappers.
    for name, func in (
        ("zlib", zlib.decompress),
        ("bz2", bz2.decompress),
        ("lzma", lzma.decompress),
    ):
        try:
            transformed = func(data)
        except Exception:
            continue
        try:
            return transformed.decode("utf-8"), [f"{name} decompression succeeded"]
        except UnicodeDecodeError:
            attempts.append(f"{name} decompression produced non-UTF-8 bytes")

    # Some payloads encode printable hex after the ascii85 layer.
    if data and all(chr(byte) in "0123456789abcdefABCDEF\r\n\t " for byte in data):
        stripped_bytes = bytearray()
        for byte in data:
            ch = chr(byte)
            if ch in "\r\n\t ":
                continue
            stripped_bytes.append(byte)
        stripped = bytes(stripped_bytes)
        if len(stripped) % 2 == 0:
            try:
                decoded = bytes.fromhex(stripped.decode("ascii"))
            except ValueError:
                attempts.append("ascii hex decode failed")
            else:
                try:
                    return decoded.decode("utf-8"), ["hex decode succeeded"]
                except UnicodeDecodeError:
                    attempts.append("hex decode produced non-UTF-8 bytes")

    try:
        return data.decode("utf-8"), ["raw UTF-8 decode succeeded"]
    except UnicodeDecodeError:
        attempts.append("raw UTF-8 decode failed")

    return None, attempts


def _format_hexdump(data: bytes, *, width: int = 16) -> str:
    lines: list[str] = []
    for offset in range(0, len(data), width):
        chunk = data[offset : offset + width]
        hex_part = " ".join(f"{byte:02x}" for byte in chunk)
        ascii_part = "".join(chr(byte) if 32 <= byte < 127 else "." for byte in chunk)
        lines.append(f"{offset:08x}  {hex_part:<{width * 3}}  {ascii_part}")
    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(description="Decode a chunk from Obfuscated.json")
    parser.add_argument("json_path", type=Path, help="Path to Obfuscated.json")
    parser.add_argument("chunk_index", type=int, help="Chunk index to decode (1-based)")
    parser.add_argument("output", type=Path, help="Where to write the decoded bytes")
    parser.add_argument("--script-key", required=True, help="Script key for XOR stage")
    parser.add_argument(
        "--bootstrapper",
        type=Path,
        help="Optional path to initv4 bootstrapper for alphabet recovery",
    )
    args = parser.parse_args()

    data = _load_json(args.json_path)
    index = args.chunk_index
    if index < 0 or index >= len(data):
        raise IndexError(f"chunk_index {index} out of bounds for {len(data)} entries")

    chunk = data[index]
    alphabet = _auto_detect_alphabet(args.bootstrapper) if args.bootstrapper else None
    decoded, strategy, text, notes = decode_chunk(
        chunk, args.script_key, alphabet=alphabet
    )

    if text is None:
        hexdump = _format_hexdump(decoded)
        attempts = ", ".join(notes) if notes else "none"
        lines = [
            f"-- chunk {index} ({len(decoded)} bytes) decoded with {strategy}; unable to recover UTF-8",
            f"-- additional attempts: {attempts}",
            "return [[",
            hexdump,
            "]]",
            "",
        ]
        args.output.write_text("\n".join(lines), encoding="utf-8")
        print(
            f"Decoded chunk {index} -> {args.output} ({len(decoded)} bytes, emitted hexdump)")
        return

    args.output.write_text(text, encoding="utf-8")
    extra = f"; {notes[0]}" if notes else ""
    print(
        f"Decoded chunk {index} -> {args.output} ({len(decoded)} bytes, UTF-8 text via {strategy}{extra})"
    )


if __name__ == "__main__":
    main()

File: lua_env/decode_chunk.lua
local json = dofile("lua_env/json.lua")
local ok_bit, bit32_lib = pcall(require, "bit32")
local bit32 = bit32_lib or bit32

local function bxor(a, b)
    if bit32 and bit32.bxor then
        return bit32.bxor(a, b)
    end
    return (a ~ b) & 0xFF
end

local function read_file(path)
    local fh, err = io.open(path, "rb")
    if not fh then
        error("failed to read file: " .. tostring(err))
    end
    local data = fh:read("*a")
    fh:close()
    return data
end

local function write_file(path, data)
    local fh, err = io.open(path, "wb")
    if not fh then
        error("failed to write file: " .. tostring(err))
    end
    fh:write(data)
    fh:close()
end

local function normalise_chunk(value)
    if type(value) == "string" then
        return value
    elseif type(value) == "table" then
        local parts = {}
        for i = 1, #value do
            parts[i] = tostring(value[i])
        end
        return table.concat(parts)
    else
        error("unsupported chunk type: " .. type(value))
    end
end

local function unique_chars(str)
    local seen = {}
    local uniq = {}
    for i = 1, #str do
        local ch = str:sub(i, i)
        if not seen[ch] then
            seen[ch] = true
            uniq[#uniq + 1] = ch
        end
    end
    return table.concat(uniq)
end

local function recover_alphabet(path)
    if not path or path == "" then
        return nil
    end
    local ok, data = pcall(read_file, path)
    if not ok then
        return nil
    end
    local best, best_len = nil, 0
    for chunk in data:gmatch("[!-~]+") do
        if #chunk >= 85 then
            local uniq = unique_chars(chunk)
            if #uniq >= 85 and #uniq > best_len then
                best = uniq
                best_len = #uniq
            end
        end
    end
    return best
end

local function ascii85_decode(input)
    local output = {}
    local values = {}
    local count = 0
    local i = 1
    while i <= #input do
        local ch = input:sub(i, i)
        if ch:match("%s") then
            -- ignore whitespace
        elseif ch == '~' then
            if input:sub(i + 1, i + 1) == '>' then
                break
            else
                error("unexpected '~' in ascii85 stream")
            end
        elseif ch == 'z' then
            if count ~= 0 then
                error("invalid ascii85 short form")
            end
            output[#output + 1] = string.char(0, 0, 0, 0)
        else
            local byte = ch:byte()
            if byte < 33 or byte > 126 then
                error(string.format("invalid ascii85 character %q (byte %d)", ch, byte))
            end
            count = count + 1
            values[count] = byte - 33
            if count == 5 then
                local accum = 0
                for j = 1, 5 do
                    accum = accum * 85 + values[j]
                end
                local b1 = math.floor(accum / 256 ^ 3) % 256
                local b2 = math.floor(accum / 256 ^ 2) % 256
                local b3 = math.floor(accum / 256) % 256
                local b4 = accum % 256
                output[#output + 1] = string.char(b1, b2, b3, b4)
                values = {}
                count = 0
            end
        end
        i = i + 1
    end
    if count > 0 then
        for j = count + 1, 5 do
            values[j] = 84
        end
        local accum = 0
        for j = 1, 5 do
            accum = accum * 85 + values[j]
        end
        local bytes = {
            math.floor(accum / 256 ^ 3) % 256,
            math.floor(accum / 256 ^ 2) % 256,
            math.floor(accum / 256) % 256,
            accum % 256,
        }
        for j = 1, count - 1 do
            output[#output + 1] = string.char(bytes[j])
        end
    end
    return table.concat(output)
end

local function z85_decode(input)
    local value = 0
    local count = 0
    local out = {}
    for i = 1, #input do
        local ch = input:sub(i, i)
        if ch:match("%s") then
            -- skip whitespace
        else
            local byte = ch:byte()
            if byte < 32 or byte > 126 then
                error(string.format("invalid z85 character %q (byte %d)", ch, byte))
            end
            value = value * 85 + (byte - 32)
            count = count + 1
            if count == 5 then
                local tmp = value
                local b4 = tmp % 256
                tmp = (tmp - b4) / 256
                local b3 = tmp % 256
                tmp = (tmp - b3) / 256
                local b2 = tmp % 256
                tmp = (tmp - b2) / 256
                local b1 = tmp % 256
                out[#out + 1] = string.char(b1, b2, b3, b4)
                value = 0
                count = 0
            end
        end
    end
    if count ~= 0 then
        error("z85 data length must be a multiple of 5")
    end
    return table.concat(out)
end

local function build_mapping(alphabet)
    local map = {}
    for i = 1, #alphabet do
        local ch = alphabet:sub(i, i)
        if map[ch] == nil then
            map[ch] = i - 1
        end
    end
    return map, #alphabet
end

local function base91_decode(input, alphabet)
    if not alphabet then
        error("base91 requires an alphabet")
    end
    local mapping, base = build_mapping(alphabet)
    local value = -1
    local buffer = 0
    local bits = 0
    local out = {}

    for i = 1, #input do
        local ch = input:sub(i, i)
        if ch:match("%s") then
            -- skip
        else
            local symbol = mapping[ch]
            if symbol == nil then
                error("invalid base91 symbol: " .. ch)
            end
            if value < 0 then
                value = symbol
            else
                local combined = value + symbol * base
                buffer = buffer + combined * 2 ^ bits
                local mod = combined % 8192
                if mod > 88 then
                    bits = bits + 13
                else
                    bits = bits + 14
                end
                while bits >= 8 do
                    local byte = buffer % 256
                    out[#out + 1] = string.char(byte)
                    buffer = (buffer - byte) / 256
                    bits = bits - 8
                end
                value = -1
            end
        end
    end

    if value >= 0 then
        buffer = buffer + value * 2 ^ bits
        local byte = buffer % 256
        out[#out + 1] = string.char(byte)
    end

    return table.concat(out)
end

local function apply_script_key(bytes, key)
    local out = {}
    local key_len = #key
    for i = 1, #bytes do
        local b = bytes:byte(i)
        if key_len > 0 then
            local k = key:byte(((i - 1) % key_len) + 1)
            b = bxor(b, k)
        end
        b = bxor(b, (i - 1) % 256)
        out[i] = string.char(b)
    end
    return table.concat(out)
end

local function is_printable(bytes)
    for i = 1, #bytes do
        local b = bytes:byte(i)
        if b == 9 or b == 10 or b == 13 then
            -- allow whitespace
        elseif b < 32 or b > 126 then
            return false
        end
    end
    return true
end

local function format_hexdump(bytes)
    local lines = {"-- binary payload; hex dump follows", "return [["}
    for offset = 0, #bytes - 1, 16 do
        local chunk = {}
        local ascii = {}
        for j = 0, 15 do
            local idx = offset + j + 1
            if idx <= #bytes then
                local b = bytes:byte(idx)
                chunk[#chunk + 1] = string.format("%02x", b)
                if b >= 32 and b < 127 then
                    ascii[#ascii + 1] = string.char(b)
                else
                    ascii[#ascii + 1] = '.'
                end
            else
                chunk[#chunk + 1] = "  "
                ascii[#ascii + 1] = ' '
            end
        end
        lines[#lines + 1] = string.format("%08x  %s  %s", offset, table.concat(chunk, " "), table.concat(ascii))
    end
    lines[#lines + 1] = "]]"
    return table.concat(lines, "\n") .. "\n"
end

local function resolve_chunk(doc, chunk_index)
    local index = chunk_index
    local first = doc[1]
    if type(first) == "table" and #first == 2 and type(first[1]) == "string" and type(first[2]) == "string" then
        index = index + 1
    end
    if index > #doc then
        error(string.format("chunk_index %d out of range after metadata adjustment (len=%d)", chunk_index, #doc))
    end
    return doc[index]
end

local function main()
    if #arg < 5 then
        io.stderr:write("usage: lua lua_env/decode_chunk.lua <Obfuscated.json> <chunk_index> <output> <script_key> <bootstrapper>\n")
        os.exit(1)
    end
    local json_path = arg[1]
    local chunk_index = tonumber(arg[2])
    local output_path = arg[3]
    local script_key = arg[4]
    local bootstrapper = arg[5]

    if not chunk_index or chunk_index < 1 then
        error("chunk_index must be >= 1")
    end

    local doc = json.decode(read_file(json_path))
    if type(doc) ~= "table" then
        error("root JSON value must be an array")
    end

    local raw_chunk = resolve_chunk(doc, chunk_index)
    local chunk = normalise_chunk(raw_chunk)
    local alphabet = recover_alphabet(bootstrapper)

    local decoded
    local ok, err = pcall(function()
        decoded = ascii85_decode(chunk)
    end)

    local method = "ascii85"
    if not ok then
        local err_msg = tostring(err)
        local tried_z85 = false
        if err_msg:find("ascii85 character", 1, true) then
            local z_ok, z_res = pcall(function()
                return z85_decode(chunk)
            end)
            if z_ok then
                decoded = z_res
                method = "z85"
                tried_z85 = true
            else
                io.stderr:write(string.format("z85 decode failed (%s)\n", tostring(z_res)))
            end
        end
        if not tried_z85 then
            if not alphabet then
                error("ascii85 failed and no alphabet available for base91: " .. err_msg)
            end
            io.stderr:write(string.format("ascii85 decode failed (%s); attempting base91\n", err_msg))
            io.stderr:write(string.format("alphabet length %d\n", #alphabet))
            decoded = base91_decode(chunk, alphabet)
            method = "base91"
        end
    end

    local transformed = apply_script_key(decoded, script_key)
    if is_printable(transformed) then
        write_file(output_path, transformed)
        print(string.format("chunk %d decoded via %s -> %s (%d bytes, printable)", chunk_index, method, output_path, #transformed))
    else
        write_file(output_path, format_hexdump(transformed))
        print(string.format("chunk %d decoded via %s -> %s (%d bytes, binary)", chunk_index, method, output_path, #transformed))
    end
end

main()
